File: apps/web/buffett_api/yahoo_finance.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

import pandas as pd
import yfinance as yf


SECTOR_ALIASES = {
    "technology": "technology",
    "healthcare": "health",
    "financial services": "finance",
    "financial": "finance",
    "industrials": "industry",
    "industrials": "industry",
    "consumer cyclical": "consumer",
    "consumer defensive": "consumer",
    "consumer staples": "consumer",
    "energy": "energy",
    "real estate": "real_estate",
    "utilities": "infrastructure",
    "communication services": "services",
    "basic materials": "industry",
}


@dataclass(frozen=True, slots=True)
class YahooSearchItem:
    symbol: str
    name: str
    exchange: str | None
    sector: str | None
    industry: str | None
    quote_type: str | None


def _extract_company_leadership(info: dict[str, Any]) -> tuple[str | None, str | None]:
    officers = info.get("companyOfficers")

    if not isinstance(officers, list):
        return None, None

    normalized_officers: list[tuple[int, str, str | None]] = []

    for officer in officers:
        if not isinstance(officer, dict):
            continue

        name = _clean_text(officer.get("name"))
        title = _clean_text(officer.get("title"))

        if not name:
            continue

        normalized_title = (title or "").lower()
        priority = 0

        if "chief executive officer" in normalized_title or normalized_title.startswith(
            "ceo"
        ):
            priority = 3
        elif " ceo" in normalized_title or "ceo " in normalized_title:
            priority = 2
        elif "chief executive" in normalized_title:
            priority = 1

        normalized_officers.append((priority, name, title))

    if not normalized_officers:
        return None, None

    normalized_officers.sort(key=lambda item: item[0], reverse=True)
    _, ceo_name, ceo_title = normalized_officers[0]
    return ceo_name, ceo_title


def _clean_text(value: Any) -> str | None:
    if value is None:
        return None

    text = str(value).strip()
    return text or None


def _clean_number(value: Any) -> float | None:
    if value is None:
        return None

    if isinstance(value, bool):
        return None

    try:
        cleaned_value = float(value)
    except (TypeError, ValueError):
        return None

    if pd.isna(cleaned_value):
        return None

    return cleaned_value


def normalize_sector(sector: str | None) -> str:
    normalized = (sector or "").strip().lower()

    if not normalized:
        return "services"

    return SECTOR_ALIASES.get(normalized, normalized.replace(" ", "_"))


def extract_first_statement_value(
    statement: pd.DataFrame,
    labels: tuple[str, ...],
) -> float | None:
    if statement.empty:
        return None

    first_column = statement.columns[0]

    for label in labels:
        if label not in statement.index:
            continue

        return _clean_number(statement.at[label, first_column])

    return None


def normalize_search_quotes(quotes: list[dict[str, Any]]) -> list[YahooSearchItem]:
    results: list[YahooSearchItem] = []

    for quote in quotes:
        if quote.get("quoteType") != "EQUITY":
            continue

        symbol = _clean_text(quote.get("symbol"))
        name = _clean_text(quote.get("longname")) or _clean_text(quote.get("shortname"))

        if not symbol or not name:
            continue

        results.append(
            YahooSearchItem(
                symbol=symbol,
                name=name,
                exchange=_clean_text(quote.get("exchDisp") or quote.get("exchange")),
                sector=_clean_text(quote.get("sector")),
                industry=_clean_text(quote.get("industry")),
                quote_type=_clean_text(quote.get("quoteType")),
            )
        )

    return results


def extract_company_profile(
    symbol: str,
    info: dict[str, Any],
    balance_sheet: pd.DataFrame,
    income_stmt: pd.DataFrame,
    cashflow: pd.DataFrame,
) -> dict[str, Any]:
    ceo_name, ceo_title = _extract_company_leadership(info)
    revenue = _clean_number(info.get("totalRevenue")) or extract_first_statement_value(
        income_stmt,
        ("Total Revenue", "Operating Revenue", "Revenue"),
    )
    net_income = extract_first_statement_value(
        income_stmt,
        (
            "Net Income",
            "Net Income Common Stockholders",
            "Net Income From Continuing And Discontinued Operation",
            "Net Income From Continuing Operation Net Minority Interest",
        ),
    )
    debt = _clean_number(info.get("totalDebt")) or extract_first_statement_value(
        balance_sheet,
        ("Total Debt", "Net Debt"),
    )
    shareholders_equity = extract_first_statement_value(
        balance_sheet,
        (
            "Common Stock Equity",
            "Stockholders Equity",
            "Total Equity Gross Minority Interest",
            "Total Equity",
        ),
    )
    free_cash_flow = _clean_number(
        info.get("freeCashflow")
    ) or extract_first_statement_value(cashflow, ("Free Cash Flow",))
    ebitda = _clean_number(info.get("ebitda")) or extract_first_statement_value(
        income_stmt,
        ("EBITDA", "Normalized EBITDA"),
    )
    pe_ratio = _clean_number(info.get("trailingPE"))
    net_margin = _clean_number(info.get("profitMargins"))
    return_on_equity = _clean_number(info.get("returnOnEquity"))

    if net_margin is not None:
        net_margin *= 100.0

    if shareholders_equity is None and return_on_equity and net_income:
        if abs(return_on_equity) > 1e-9:
            shareholders_equity = net_income / return_on_equity

    return {
        "symbol": symbol.upper(),
        "name": _clean_text(info.get("longName")) or _clean_text(info.get("shortName")) or symbol.upper(),
        "sector": normalize_sector(_clean_text(info.get("sector"))),
        "industry": _clean_text(info.get("industry")),
        "market_price": _clean_number(info.get("regularMarketPrice"))
        or _clean_number(info.get("currentPrice")),
        "country": _clean_text(info.get("country")),
        "exchange": _clean_text(info.get("exchange")) or _clean_text(info.get("fullExchangeName")),
        "currency": _clean_text(info.get("currency")),
        "website": _clean_text(info.get("website")),
        "city": _clean_text(info.get("city")),
        "state": _clean_text(info.get("state")),
        "full_time_employees": _clean_number(info.get("fullTimeEmployees")),
        "ceo_name": ceo_name,
        "ceo_title": ceo_title,
        "summary": _clean_text(info.get("longBusinessSummary")),
        "revenue": revenue,
        "net_income": net_income,
        "debt": debt,
        "shareholders_equity": shareholders_equity,
        "free_cash_flow": free_cash_flow,
        "ebitda": ebitda,
        "pe_ratio": pe_ratio,
        "net_margin": net_margin,
        "environmental_score": 60.0,
        "social_score": 60.0,
        "governance_score": 60.0,
        "source": "yahoo_finance",
    }


class YahooFinanceService:
    def search_companies(self, query: str) -> list[dict[str, Any]]:
        sanitized_query = query.strip()

        if len(sanitized_query) < 2:
            return []

        search = yf.Search(
            sanitized_query,
            max_results=8,
            news_count=0,
            lists_count=0,
            include_nav_links=False,
            include_research=False,
            raise_errors=True,
        )

        return [
            {
                "symbol": result.symbol,
                "name": result.name,
                "exchange": result.exchange,
                "sector": result.sector,
                "industry": result.industry,
                "quote_type": result.quote_type,
            }
            for result in normalize_search_quotes(search.quotes or [])
        ]

    def get_company_profile(self, symbol: str) -> dict[str, Any]:
        sanitized_symbol = symbol.strip().upper()

        if not sanitized_symbol:
            raise ValueError("Le ticker Yahoo Finance est obligatoire.")

        ticker = yf.Ticker(sanitized_symbol)
        return extract_company_profile(
            sanitized_symbol,
            ticker.info or {},
            ticker.balance_sheet,
            ticker.income_stmt,
            ticker.cashflow,
        )

File: apps/web/src/lib/yahoo-finance.ts
import type {
  BuffettCompanyLookupResult,
  BuffettSearchSuggestion,
} from "@/lib/buffett-api";

const YAHOO_HEADERS = {
  "User-Agent": "Mozilla/5.0",
  Accept: "application/json",
};

type YahooSearchApiResponse = {
  quotes?: Array<Record<string, unknown>>;
};

type YahooTimeseriesApiResponse = {
  timeseries?: {
    result?: Array<Record<string, unknown>>;
  };
};

type YahooQuoteApiResponse = {
  quoteResponse?: {
    result?: Array<Record<string, unknown>>;
  };
};

function normalizeText(value: unknown) {
  return typeof value === "string" ? value.trim() : "";
}

function parseNumber(value: unknown) {
  return typeof value === "number" && Number.isFinite(value) ? value : null;
}

function normalizeSector(sector: string | null | undefined) {
  const normalizedSector = (sector ?? "").trim().toLowerCase();

  if (!normalizedSector) {
    return "services";
  }

  switch (normalizedSector) {
    case "technology":
      return "technology";
    case "healthcare":
      return "health";
    case "financial services":
    case "financial":
      return "finance";
    case "industrials":
    case "basic materials":
      return "industry";
    case "consumer cyclical":
    case "consumer defensive":
    case "consumer staples":
      return "consumer";
    case "energy":
      return "energy";
    case "real estate":
      return "real_estate";
    case "utilities":
      return "infrastructure";
    case "communication services":
      return "services";
    default:
      return normalizedSector.replace(/\s+/g, "_");
  }
}

function extractLatestTimeseriesValue(
  seriesEntries: Array<Record<string, unknown>> | undefined
) {
  if (!Array.isArray(seriesEntries) || seriesEntries.length === 0) {
    return null;
  }

  const latestEntry = [...seriesEntries]
    .filter((entry) => typeof entry.asOfDate === "string")
    .sort((leftEntry, rightEntry) =>
      String(rightEntry.asOfDate).localeCompare(String(leftEntry.asOfDate))
    )[0];

  if (!latestEntry || typeof latestEntry !== "object") {
    return null;
  }

  const reportedValue =
    latestEntry.reportedValue &&
    typeof latestEntry.reportedValue === "object" &&
    "raw" in latestEntry.reportedValue
      ? latestEntry.reportedValue.raw
      : null;

  return parseNumber(reportedValue);
}

function findTimeseriesValue(
  payload: YahooTimeseriesApiResponse,
  key: string
) {
  const result = payload.timeseries?.result?.find(
    (entry) =>
      Array.isArray(entry.meta) === false &&
      entry.meta &&
      typeof entry.meta === "object" &&
      "type" in entry.meta &&
      Array.isArray((entry.meta as { type?: unknown }).type) &&
      (entry.meta as { type: unknown[] }).type[0] === key
  );

  if (!result) {
    return null;
  }

  const seriesEntries = result[key];
  return extractLatestTimeseriesValue(
    Array.isArray(seriesEntries)
      ? (seriesEntries as Array<Record<string, unknown>>)
      : undefined
  );
}

export async function searchYahooFinanceCompanies(
  query: string
): Promise<BuffettSearchSuggestion[]> {
  const sanitizedQuery = query.trim();

  if (sanitizedQuery.length < 2) {
    return [];
  }

  const response = await fetch(
    `https://query1.finance.yahoo.com/v1/finance/search?q=${encodeURIComponent(
      sanitizedQuery
    )}&quotesCount=8&newsCount=0&listsCount=0&enableFuzzyQuery=true`,
    {
      headers: YAHOO_HEADERS,
      cache: "no-store",
    }
  );

  if (!response.ok) {
    throw new Error(`Yahoo search failed with status ${response.status}`);
  }

  const payload = (await response.json()) as YahooSearchApiResponse;

  return (payload.quotes ?? [])
    .filter((quote) => normalizeText(quote.quoteType) === "EQUITY")
    .map((quote) => {
      const symbol = normalizeText(quote.symbol);
      const name = normalizeText(quote.longname) || normalizeText(quote.shortname);

      if (!symbol || !name) {
        return null;
      }

      return {
        symbol,
        name,
        exchange: normalizeText(quote.exchDisp) || normalizeText(quote.exchange) || null,
        sector: normalizeText(quote.sector) || null,
        industry: normalizeText(quote.industry) || null,
        quoteType: normalizeText(quote.quoteType) || null,
      } satisfies BuffettSearchSuggestion;
    })
    .filter((quote): quote is BuffettSearchSuggestion => quote !== null);
}

export async function getYahooFinanceCompanyProfile(
  symbol: string
): Promise<BuffettCompanyLookupResult> {
  const sanitizedSymbol = symbol.trim().toUpperCase();

  if (!sanitizedSymbol) {
    throw new Error("Le ticker Yahoo Finance est obligatoire.");
  }

  const searchRequest = searchYahooFinanceCompanies(sanitizedSymbol);
  const fundamentalsRequest = fetch(
    `https://query1.finance.yahoo.com/ws/fundamentals-timeseries/v1/finance/timeseries/${encodeURIComponent(
      sanitizedSymbol
    )}?type=annualTotalRevenue,annualNetIncomeContinuousOperations,annualFreeCashFlow,annualTotalDebt,annualStockholdersEquity,annualEBITDA&period1=1609459200&period2=1893456000`,
    {
      headers: YAHOO_HEADERS,
      cache: "no-store",
    }
  );
  const quoteRequest = Promise.resolve(
    fetch(
      `https://query1.finance.yahoo.com/v7/finance/quote?symbols=${encodeURIComponent(
        sanitizedSymbol
      )}`,
      {
        headers: YAHOO_HEADERS,
        cache: "no-store",
      }
    )
  ).catch(() => null);

  const [searchResults, fundamentalsResponse, quoteResponse] = await Promise.all([
    searchRequest,
    fundamentalsRequest,
    quoteRequest,
  ]);

  if (!fundamentalsResponse.ok) {
    throw new Error(
      `Yahoo fundamentals failed with status ${fundamentalsResponse.status}`
    );
  }

  const metadata =
    searchResults.find((company) => company.symbol === sanitizedSymbol) ??
    searchResults[0] ??
    null;

  const fundamentalsPayload =
    (await fundamentalsResponse.json()) as YahooTimeseriesApiResponse;
  const quotePayload = quoteResponse && quoteResponse.ok
    ? ((await quoteResponse.json()) as YahooQuoteApiResponse)
    : null;
  const quote =
    quotePayload?.quoteResponse?.result?.find(
      (entry) => normalizeText(entry.symbol) === sanitizedSymbol
    ) ?? quotePayload?.quoteResponse?.result?.[0] ?? null;

  const revenue = findTimeseriesValue(fundamentalsPayload, "annualTotalRevenue");
  const netIncome = findTimeseriesValue(
    fundamentalsPayload,
    "annualNetIncomeContinuousOperations"
  );
  const debt = findTimeseriesValue(fundamentalsPayload, "annualTotalDebt");
  const shareholdersEquity = findTimeseriesValue(
    fundamentalsPayload,
    "annualStockholdersEquity"
  );
  const freeCashFlow = findTimeseriesValue(
    fundamentalsPayload,
    "annualFreeCashFlow"
  );
  const ebitda = findTimeseriesValue(fundamentalsPayload, "annualEBITDA");
  const netMargin =
    revenue && netIncome && revenue !== 0 ? (netIncome / revenue) * 100 : null;

  return {
    symbol: sanitizedSymbol,
    name: metadata?.name ?? sanitizedSymbol,
    sector: normalizeSector(metadata?.sector),
    industry: metadata?.industry ?? null,
    marketPrice: parseNumber(quote?.regularMarketPrice),
    country: null,
    exchange: metadata?.exchange ?? null,
    currency: normalizeText(quote?.currency) || null,
    website: null,
    city: null,
    state: null,
    fullTimeEmployees: null,
    ceoName: null,
    ceoTitle: null,
    summary: null,
    revenue,
    netIncome,
    debt,
    shareholdersEquity,
    freeCashFlow,
    ebitda,
    peRatio: 15,
    netMargin,
    environmentalScore: 60,
    socialScore: 60,
    governanceScore: 60,
    source: "yahoo_finance_public",
  };
}
